### PR TITLE
S3 client side encryption

### PR DESCRIPTION
**Pull request by xuzha**
Closes https://github.com/elastic/elasticsearch/issues/13673

I'm picking up a unfinished PR here https://github.com/elastic/elasticsearch-cloud-aws/pull/118. Most of the implementation keeps the same.

This PR add s3 client side encryption functionality. It uses the envelope encryption of the AWS ADK for Java (AES 256 in CBC mode). You case use keys of 128, 192 or 256 bits but you have to install the JCE because the envelope encryption will use 256bits keys.

I'm not sure about this: I remove the MD5 checking part in the `doUpload` method. I think AWS S3 SDK is doing this for us. 
More here: https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/AmazonS3Client.java#L1449-L1469.

### CODE REVIEW

**Comment by xuzha** [comment]
@dadoonet would you like to take a look, if you have time? :-)

File: plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/AwsS3Service.java [hunk 1]
```diff
@@ -147,4 +148,5 @@
     }
 
     AmazonS3 client(String endpoint, Protocol protocol, String region, String account, String key, Integer maxRetries);
+    AmazonS3 client(String endpoint, Protocol protocol, String region, String account, String key, Integer maxRetries, EncryptionMaterials clientSideEncryptionMaterials);
```
Comment (dadoonet) [inline plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/AwsS3Service.java]: I think we should only have one client method here. So I would replace the previous line by the new method.

File: plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/InternalAwsS3Service.java [hunk 2]
```diff
@@ -49,7 +54,7 @@
     /**
      * (acceskey, endpoint) -&gt; client
      */
-    private Map<Tuple<String, String>, AmazonS3Client> clients = new HashMap<>();
+    private Map<Tuple<String, Tuple<String, EncryptionMaterials>>, AmazonS3Client> clients = new HashMap<>();
```
Comment (dadoonet) [inline plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/InternalAwsS3Service.java]: I don't understand why we need to add the `EncryptionMaterials` here. Does it mean that we want to have many client instances using the same account but with different encryption settings?

File: plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/InternalAwsS3Service.java [hunk 3]
```diff
@@ -123,7 +136,19 @@ private synchronized AmazonS3 getClient(String endpoint, Protocol protocol, Stri
...
+
+        if (clientSideEncryptionMaterials != null) {
+            EncryptionMaterialsProvider encryptionMaterialsProvider = new StaticEncryptionMaterialsProvider(clientSideEncryptionMaterials);
+            CryptoConfiguration cryptoConfiguration = new CryptoConfiguration();
+            client = new AmazonS3EncryptionClient(
+                    credentials,
+                    encryptionMaterialsProvider,
+                    clientConfiguration,
+                    cryptoConfiguration
+            );
```
Comment (dadoonet) [inline plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/InternalAwsS3Service.java]: Small formatting comment. Can you put this `);` at the end of the previous line?

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 4]
```diff
@@ -252,14 +296,30 @@ public S3Repository(RepositoryName name, RepositorySettings repositorySettings,
         String storageClass = getValue(repositorySettings, Repository.STORAGE_CLASS_SETTING, Repositories.STORAGE_CLASS_SETTING);
         String cannedACL = getValue(repositorySettings, Repository.CANNED_ACL_SETTING, Repositories.CANNED_ACL_SETTING);
 
-        logger.debug("using bucket [{}], region [{}], endpoint [{}], protocol [{}], chunk_size [{}], server_side_encryption [{}], buffer_size [{}], max_retries [{}], cannedACL [{}], storageClass [{}]",
+        logger.debug("using bqucket [{}], region [{}], endpoint [{}], protocol [{}], chunk_size [{}], server_side_encryption [{}], buffer_size [{}], max_retries [{}], cannedACL [{}], storageClass [{}]",
```
Comment (dadoonet) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: s/bqucket/bucket ?

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 5]
```diff
@@ -252,14 +296,30 @@ public S3Repository(RepositoryName name, RepositorySettings repositorySettings,
...
                 bucket, region, endpoint, protocol, chunkSize, serverSideEncryption, bufferSize, maxRetries, cannedACL, storageClass);
 
         String key = getValue(repositorySettings, Repository.KEY_SETTING, Repositories.KEY_SETTING);
         String secret = getValue(repositorySettings, Repository.SECRET_SETTING, Repositories.SECRET_SETTING);
 
-        blobStore = new S3BlobStore(settings, s3Service.client(endpoint, protocol, region, key, secret, maxRetries),
-                bucket, region, serverSideEncryption, bufferSize, maxRetries, cannedACL, storageClass);
+        // parse and validate the client side encryption setting
+        String symmetricKeyBase64 = getValue(repositorySettings, Repository.CLIENT_SYMMETRIC_KEY, Repositories.CLIENT_SYMMETRIC_KEY);
+        String publicKeyBase64 =getValue(repositorySettings, Repository.CLIENT_PUBLIC_KEY, Repositories.CLIENT_PUBLIC_KEY);
```
Comment (dadoonet) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: Missing space. Should be: `String publicKeyBase64 = getValue`

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 6]
```diff
@@ -271,6 +331,52 @@ public S3Repository(RepositoryName name, RepositorySettings repositorySettings,
...
+
+    /**
+     * Init and verify initClientSideEncryption settings
+     */
+    private EncryptionMaterials initClientSideEncryption(String symmetricKey, String publicKey, String privateKey, RepositoryName name) {
+
+        EncryptionMaterials clientSideEncryptionMaterials = null;
+
+        if (Strings.isNullOrEmpty(symmetricKey) == false && (Strings.isNullOrEmpty(publicKey) == false || Strings.isNullOrEmpty(privateKey) == false)) {
+            throw new RepositoryException(name.name(), "Client-side encryption: You can't specify an symmetric key AND a public/private key pair");
```
Comment (dadoonet) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: You can't specify **a** symmetric key ?

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 7]
```diff
@@ -271,6 +331,52 @@ public S3Repository(RepositoryName name, RepositorySettings repositorySettings,
...
+                // Check crypto
+                if (Cipher.getMaxAllowedKeyLength("AES") < 256) {
+                    throw new RepositoryException(name.name(), "Client-side encryption: Please install the Java Cryptography Extension");
+                }
+
+                // Transform the keys in a EncryptionMaterials
+                if (Strings.isNullOrEmpty(symmetricKey) == false) {
+                    clientSideEncryptionMaterials = new EncryptionMaterials(new SecretKeySpec(Base64.decode(symmetricKey), "AES"));
+                } else {
+                    if (Strings.isNullOrEmpty(publicKey)|| Strings.isNullOrEmpty(privateKey)){
```
Comment (dadoonet) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: s/`Strings.isNullOrEmpty(publicKey)||`/`Strings.isNullOrEmpty(publicKey) ||`

(space missing)

**Comment by dadoonet** [comment]
I gave a first look (not complete yet) and it's a really good start.

About removing MD5 check, I do agree that it sounds already here in the AWS SDK. Any chance you can test manually that it goes through this line in debug mode? https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/AmazonS3Client.java#L1455

File: plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/InternalAwsS3Service.java (hunk 2 above)
Comment (xuzha) [inline plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/InternalAwsS3Service.java]: Thx David, I was confused too, I think this is make sense, this allows you to use a different key for each snapshot repository.

For example you can have two repos, with two different keys : 
```
PUT _snapshot/test-1
{
  "type": "s3",
  "settings": {
    "bucket": "xu-client-1",
    "region": "us-west-2",
    "client_symmetric_key":"Hk/RABbqRnhgSAMXiVUV0w=="
  }
}

```

```
PUT _snapshot/test-2
{
  "type": "s3",
  "settings": {
    "bucket": "xu-client-2",
    "region": "us-west-2",
    "client_public_key":"MFwwDQYJKoZIhvcNAQEBBQADSwAwSAJBALb4IRgaDtmwsz+kC/bkKLno5bRWMDbRDztcM/N/VJIg+HQKA8ees6uznEOGe6cOZp+QHYrpTIXs36QB9vfsVdUCAwEAAQ==",
    "client_private_key":"MIIBUwIBADANBgkqhkiG9w0BAQEFAASCAT0wggE5AgEAAkEAtvghGBoO2bCzP6QL9uQouejltFYwNtEPO1wz839UkiD4dAoDx56zq7OcQ4Z7pw5mn5AdiulMhezfpAH29+xV1QIDAQABAkBEy/OVnmarD7e2XDZrdMqjbKDCOA4U7nKtvTODgQMJll0b7wA52bY/kG8gA8Q2aqiKVHDRl/EQ33bELr2A56NBAiEA4+7qC/TMw2V2q+FuW5Az8mzfCH6mNmyMLc/XwuZ2/V0CIQDNf9NRbZ0EZQCiq6iBjgYzkdJlv1tDt9erqVj0w0K62QIgZDE5IFhTSfDn4VYOtKEGtKG2yH0jgvjkBZ8/MKUt2OECIFSzDuJND560EqL5paZgZ2XyAIo3aOJsb9QtJKEdqe9hAiBgithEZGGQNV/pXweOtku/CezLvY2FJaSPShfeQfXjcg=="
  }
}

```

**Comment by xuzha** [comment]
Thx @dadoonet, I just updated the PR ;-)

About the debugging, I confirmed that AWS SDK is checking the MD5 for us, the screen shot: 
screen shot 2016-03-05 at 14 48 48

**Comment by xuzha** [comment]
@dadoonet can you take another look when you have time?

File: docs/plugins/repository-s3.asciidoc [hunk 8]
```diff
@@ -236,6 +236,16 @@ The following settings are supported:
...
+`client_symmetric_key`::
+    Sets the keys to use to encrypt your snapshots. You can specify either a symmetric key or a public/private key pair.
+    No encryption by default. This sets a Base64-encoded AES symmetric-key (128, 192 or 256 bits)
+
+`client_public_key`::
+    Sets the a base64-encoded RSA public key
+
+`client_private_key`::
+    Sets the a base64-encoded RSA private key
+
```
Comment (dadoonet) [inline docs/plugins/repository-s3.asciidoc]: I was wondering about this. It's all about encryption.

We already have `server_side_encryption`. I wonder if we should at some point rename the settings (may be within another PR) to:

```js
PUT _snapshot/my_s3_repository
{
  "type": "s3",
  "settings": {
    "bucket": "my_bucket_name",
    "region": "us-west",
    "encryption": {
       "server": {
           "enabled":  false
       },
       "client": {
           "enabled":  true,
           "symmetric_key": "XYZ",
           "public_key": "XYZ",
           "private_key": "XYZ"
       }
    }
  }
}
```

Just thinking out loud here. Might be too much engineering though...

File: plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/InternalAwsS3Service.java (hunk 2 above)
Comment (dadoonet) [inline plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/InternalAwsS3Service.java]: Ok that makes sense to me.

**Comment by dadoonet** [comment]
I left a note which we can discuss in another issuer later. The current PR looks good to me.

**Comment by xuzha** [comment]
Merged into master. Thanks @dadoonet  for the review and sorry for the delay.

**Comment by xuzha** [comment]
@clintongormley I have noticed @NicolasTr has not signed the CLA. Is there any problem here?

**Comment by GlenRSmith** [comment]
@xuzha if it helps, @NicolasTr signed the CLA in:
https://github.com/elastic/elasticsearch-cloud-aws/pull/118

**Comment by jasontedor** [comment]
I don't think that this pull request should have been merged. The CLA doesn't pass on one of the commits, and one of the [commits](https://github.com/elastic/elasticsearch/pull/16843/commits/ea78fd6560624968fada645c6358eea84e4afb7f#diff-9df9f5bfe09e4d36bf31d31c387a3c94R22) has merge conflict markers in it.

**Comment by xuzha** [comment]
Thanks @jasontedor 

The first two commits should be together, I picked up a really old commit from elastic/elasticsearch-cloud-aws#118, and fixed in the later commit.

**Comment by jasontedor** [comment]
> The first two commits should be together, I picked up a really old commit from [elastic/elasticsearch-cloud-aws#118](https://github.com/elastic/elasticsearch-cloud-aws/pull/118), and fixed in the later commit.

We should never knowingly commit code that doesn't compile.

**Comment by billxinli** [comment]
Curious if we will ever see this feature implemented or merged in?

**Comment by hoffoo** [comment]
Any hope that we will see this?

**Comment by dchang-novotec** [comment]
^^^ ditto, would definitely like to see this feature in the next release

**Comment by stantona** [comment]
It would be great to see this feature merged.

**Comment by sts** [comment]
Any news on this?

**Comment by sts** [comment]
Any ideas whats the status of this?

**Comment by kogelc** [comment]
Hello, do you know when this feature will be added ?

**Comment by JigarJoshi** [comment]
🏓  Any update here ?
